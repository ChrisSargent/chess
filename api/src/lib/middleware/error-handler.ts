/* eslint-disable no-param-reassign */
import middy from "@middy/core";
import { AxiosError } from "axios";
import HttpStatus from "http-status-codes";
import { ErrorResponse } from "../types";

const removeUndefined = (obj: any) =>
  Object.entries(obj).reduce((a, [k, v]) => (v === undefined ? a : { ...a, [k]: v }), {});

// A generic server error
const serverErrResponse: ErrorResponse = {
  source: "graph",
  status: HttpStatus.INTERNAL_SERVER_ERROR,
  title: HttpStatus.getStatusText(HttpStatus.INTERNAL_SERVER_ERROR),
};
const gwTOutErrResponse: ErrorResponse = {
  source: "graph",
  status: HttpStatus.GATEWAY_TIMEOUT,
  title: HttpStatus.getStatusText(HttpStatus.GATEWAY_TIMEOUT),
};

export const errorHandler = (logger: any) => ({
  onError: (handler: middy.HandlerLambda, next: middy.NextFunction) => {
    logger.debug(handler);
    if (!(handler.error instanceof Error)) {
      return next();
    }
    try {
      const error = handler.error as AxiosError;

      // The request was made and the server responded with a status code that falls out of the range of 2xx
      if (error.response) {
        logger.createSubLogger("gateway_error").error(error.response.data);
        handler.response = error.response.data;
        return next();
      }

      // The request was made but no response was received
      if (error.request) {
        const { baseURL, data, headers, method, params } = error.config;
        const logMe = removeUndefined({
          baseURL,
          data,
          headers,
          method,
          params,
        });
        logger.createSubLogger("gateway_timeout").error(logMe);

        handler.response = { error: gwTOutErrResponse };
        return next();
      }
      // Something unknown happened during function invocation that triggered an Error
      logger.createSubLogger("function_error").error(handler.error.message);

      // If response.error has already been set, leave as is
      if (handler.response && handler.response.error) {
        return next();
      }

      handler.response = { error: serverErrResponse };
      return next();
    } catch (err) {
      // Something happened in this function that caused an error
      logger.createSubLogger("caught_error").error(err);

      handler.response = { error: serverErrResponse };
      return next();
    }
  },
});
