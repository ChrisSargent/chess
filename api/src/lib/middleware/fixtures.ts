import createError from "axios/lib/core/createError";

export const config = {
  baseURL: "http://localhost:8081",
  headers: {
    Accept: "application/json, text/plain, */*",
  },
  method: "get",
  params: { objectType: "orderBook", userID: "investor1" },
  url: "/partial",
};

export const request = {
  method: "GET",
  noshow: "remove this field",
  path: "/partial?objectType=orderBook&userID=investor1",
};

export const response = {
  data: {
    error: {
      status: 401,
      title: "User authentication failed: user not found",
    },
  },
  headers: {
    "content-type": "application/problem+json; charset=utf-8",
    "x-content-type-options": "nosniff",
  },
  status: 401,
  statusText: "Unauthorized",
};

export const axios401 = createError(null, config, null, request, response);
export const axiosNoResponse = createError(null, config, null, request, null);
export const unknownErr = new Error("UNKNOWN ERROR");
export const unAuthResp = {
  error: {
    detail: "No identity",
    status: 401,
    title: "Unauthorised",
  },
};
