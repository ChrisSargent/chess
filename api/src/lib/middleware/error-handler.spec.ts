import middy from "@middy/core";
import { AxiosError } from "axios";
import { errorHandler } from "./error-handler";
import * as fixtures from "./fixtures";

describe("Error Handling Middleware", () => {
  const defHandler: middy.HandlerLambda = {
    callback: jest.fn(),
    context: {
      awsRequestId: "awsIdxxx",
      callbackWaitsForEmptyEventLoop: true,
      done: jest.fn(),
      fail: jest.fn(),
      functionName: "TestFn",
      functionVersion: "1.0",
      getRemainingTimeInMillis: jest.fn(),
      invokedFunctionArn: "arnxxx",
      logGroupName: "lgnxxx",
      logStreamName: "lsnxxx",
      memoryLimitInMB: "128",
      succeed: jest.fn(),
    },
    error: {
      message: "",
      name: "",
    },
    event: {},
    response: "Original Response",
  };

  const next = jest.fn();

  const sublogger = {
    createSubLogger: jest.fn(),
    debug: jest.fn(),
    error: jest.fn(),
  };

  const logger = {
    createSubLogger: jest.fn(() => sublogger),
    debug: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("Doesn't alter the function result if there is no error", () => {
    const handler: middy.HandlerLambda = { ...defHandler };
    errorHandler(logger).onError(handler, next);

    expect(handler.response).toMatchSnapshot();
    expect(logger.createSubLogger).toHaveBeenCalledTimes(0);
    expect(sublogger.error).toHaveBeenCalledTimes(0);
    expect(next).toHaveBeenCalledTimes(1);
  });

  it("Returns the error response data from a completed network call", () => {
    const handler: middy.HandlerLambda = { ...defHandler, error: fixtures.axios401 };
    errorHandler(logger).onError(handler, next);

    expect(next).toHaveBeenCalledTimes(1);
    expect(handler.response).toMatchSnapshot();
    expect(logger.createSubLogger).toHaveBeenCalledTimes(1);
    expect(logger.createSubLogger).toHaveBeenCalledWith("gateway_error");

    const error = handler.error as AxiosError;
    expect(sublogger.error).toHaveBeenCalledWith(error.response!.data);
  });

  it("Returns gateway timeout message from an uncomplete network call", () => {
    const handler: middy.HandlerLambda = { ...defHandler, error: fixtures.axiosNoResponse };
    errorHandler(logger).onError(handler, next);

    expect(next).toHaveBeenCalledTimes(1);
    expect(handler.response).toMatchSnapshot();
    expect(logger.createSubLogger).toHaveBeenCalledTimes(1);
    expect(logger.createSubLogger).toHaveBeenCalledWith("gateway_timeout");

    const error = handler.error as AxiosError;
    const { baseURL, data, headers, method, params } = error.config;
    expect(sublogger.error).toHaveBeenCalledWith({
      baseURL,
      data,
      headers,
      method,
      params,
    });
  });

  it("Returns generic error message if there's an unknown code error", () => {
    const handler: middy.HandlerLambda = { ...defHandler, error: fixtures.unknownErr };
    errorHandler(logger).onError(handler, next);

    expect(next).toHaveBeenCalledTimes(1);
    expect(handler.response).toMatchSnapshot();
    expect(logger.createSubLogger).toHaveBeenCalledTimes(1);
    expect(logger.createSubLogger).toHaveBeenCalledWith("function_error");
    expect(sublogger.error).toHaveBeenCalledWith(fixtures.unknownErr.message);
  });

  it("Returns response as is, if included", () => {
    const handler: middy.HandlerLambda = {
      ...defHandler,
      error: fixtures.unknownErr,
      response: fixtures.unAuthResp,
    };
    errorHandler(logger).onError(handler, next);

    expect(handler.response).toMatchSnapshot();
    expect(next).toHaveBeenCalledTimes(1);
    expect(logger.createSubLogger).toHaveBeenCalledTimes(1);
    expect(logger.createSubLogger).toHaveBeenCalledWith("function_error");
    expect(sublogger.error).toHaveBeenCalledWith(fixtures.unknownErr.message);
  });

  it("Returns response in case of an error inside the error handler - the catch", () => {
    // Cause the try/catch in the function to throw an error
    const next2 = jest.fn(() => {
      throw new Error("Shoud be in Catch");
    });
    const handler: middy.HandlerLambda = {
      ...defHandler,
      error: fixtures.unknownErr,
    };

    // Ensures the assertions in the try/catch to ensure we're hitting it
    expect.hasAssertions();
    try {
      errorHandler(logger).onError(handler, next2);
    } catch (error) {
      expect(handler.response).toMatchSnapshot();
      // Will be called twice because of the way we're causing the error
      expect(next2).toHaveBeenCalledTimes(2);
      expect(logger.createSubLogger).toHaveBeenCalledTimes(2);
      expect(logger.createSubLogger).toHaveBeenCalledWith("caught_error");
      expect(sublogger.error).toHaveBeenCalledWith(fixtures.unknownErr.message);
    }
  });
});
