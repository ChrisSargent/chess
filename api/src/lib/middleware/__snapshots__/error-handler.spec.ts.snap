// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Handling Middleware Doesn't alter the function result if there is no error 1`] = `"Original Response"`;

exports[`Error Handling Middleware Returns gateway timeout message from an uncomplete network call 1`] = `
Object {
  "error": Object {
    "source": "blockgraph",
    "status": 504,
    "title": "Gateway Timeout",
  },
}
`;

exports[`Error Handling Middleware Returns generic error message if there's an unknown code error 1`] = `
Object {
  "error": Object {
    "source": "blockgraph",
    "status": 500,
    "title": "Server Error",
  },
}
`;

exports[`Error Handling Middleware Returns response as is, if included 1`] = `
Object {
  "error": Object {
    "detail": "No identity",
    "status": 401,
    "title": "Unauthorised",
  },
}
`;

exports[`Error Handling Middleware Returns response in case of an error inside the error handler - the catch 1`] = `
Object {
  "error": Object {
    "source": "blockgraph",
    "status": 500,
    "title": "Server Error",
  },
}
`;

exports[`Error Handling Middleware Returns the error response data from a completed network call 1`] = `
Object {
  "error": Object {
    "status": 401,
    "title": "User authentication failed: user not found",
  },
}
`;
